### ISSUE TITLE
Refactor Source building

### ISSUE THREAD

**Issue by fulmicoton** [quickwit-inc/quickwit#1064]
In #1058, we did one step from `(dynamic factory + json)` to a `static enum` world in order to have source type specific source params validation.

We are now half-way between two worlds.

# Solution 1

Just go full enum.
The factory is a function of the enum.
Pros: Very simple
Cons: Does not separate the concern super well.

# Solution 2

Use a SourceParams trait that can build a Source type (does the job of the factory) and do dark magic, like what @fmassot did for the IndexConfig deserialization.

**Comment by fulmicoton** [comment]
@fmassot @guilload what do you think?

**Comment by fmassot** [comment]
yes totally! I must say that @guilload stopped himself in the middle of the river to have something ready for 0.2 :)

**Comment by fulmicoton** [comment]
Solution 1 or solution 2?

**Comment by guilload** [comment]
I was working around that code yesterday, and I'm leaning towards solution 1. `source_type()` does now return an enum. We can dispatch directly and statically in `load_source`. We can optionally keep the `SourceFactory` trait around to preserve the decoupling.
